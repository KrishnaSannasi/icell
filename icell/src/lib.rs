#![cfg_attr(not(feature = "std"), no_std)]

#[macro_export]
macro_rules! write_all {
    ($owner:expr => $($cells:expr),*) => {{
        $crate::write_all!(@destruct [()] [$(($cells))*] $owner.write_all($crate::hlist!($(&$cells as &$crate::ICell<_, _>),*)))
    }};
    (@destruct [$($rest:tt)*] [$first:tt $($cells:tt)*] $value:expr) => {
        match $value {
            $crate::hlist::Cons(value, rest) => $crate::write_all!(@destruct [$($rest)* , value] [$($cells)*] rest)
        }
    };
    (@destruct [()] [] $value:expr) => {
        match $value { $crate::hlist::Nil => () }
    };
    (@destruct [(), $($rest:tt)*] [] $value:expr) => {
        match $value { $crate::hlist::Nil => ($($rest)*) }
    };
}

mod core;
pub use self::core::{ICell, Identifier, Owner, Transparent};

#[doc(hidden)]
pub mod hlist;

pub mod generative;
pub mod immovable;
pub mod runtime;
pub mod typeid;
#[cfg(feature = "std")]
pub mod typeid_tl;
