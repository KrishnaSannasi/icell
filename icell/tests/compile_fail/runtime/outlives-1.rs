use icell::runtime_id;

runtime_id!(type Runtime(usize));

fn main() {
    let owner = Runtime::owner();
    let value = owner.cell(10);

    let x = owner.read(&value);

    drop(owner);

    assert_eq!(*x, 10);
}
