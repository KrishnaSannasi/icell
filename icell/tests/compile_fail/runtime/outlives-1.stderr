error[E0505]: cannot move out of `owner` because it is borrowed
  --> $DIR/outlives-1.rs:11:10
   |
9  |     let x = owner.read(&value);
   |             ----- borrow of `owner` occurs here
10 | 
11 |     drop(owner);
   |          ^^^^^ move out of `owner` occurs here
12 | 
13 |     assert_eq!(*x, 10);
   |     ------------------- borrow later used here
